File: js-automation/pages/LoginPage.js
import { By } from 'selenium-webdriver';
import { BasePage } from './BasePage.js';

export class LoginPage extends BasePage {
  constructor(driver, baseUrl) {
    super(driver, baseUrl);
    this.emailInput = By.css("[data-testid='login-email']");
    this.passwordInput = By.css("[data-testid='login-password']");
    this.submitButton = By.css("[data-testid='login-submit']");
    this.closeButton = By.css("[data-testid='login-modal-close']");
    this.userMenuButton = By.css("[data-testid='user-menu-btn']");
    this.signOutButton = By.css("[data-testid='sign-out-btn']");
    this.errorAlert = By.css("[data-testid='login-error']");
    this.toggleRegisterButton = By.css("[data-testid='toggle-register-btn']");
    this.toggleLoginButton = By.css("[data-testid='toggle-login-btn']");
    this.registerNameInput = By.css("[data-testid='register-name']");
  }

  async openModal() {
    const isModalOpen = await this.isElementPresent(this.emailInput, 1000);
    if (!isModalOpen) {
      const loggedIn = await this.isElementPresent(this.userMenuButton, 1000);
      if (loggedIn) {
        await this.logout();
      }
      const signInBtn = By.css("[data-testid='signin-button'], #navbar-signin-button");
      await this.click(signInBtn);
    }
    await this.waitForVisible(this.emailInput);
  }

  async login(email, password) {
    await this.openModal();

    // Ensure we are on login view if register is active
    const hasToggleLogin = await this.isElementPresent(this.toggleLoginButton, 500);
    if (hasToggleLogin) {
      await this.click(this.toggleLoginButton);
      await this.sleep(200);
    }

    await this.type(this.emailInput, email);
    await this.type(this.passwordInput, password);
    await this.click(this.submitButton);
    await this.sleep(600);
  }

  async register(fullName, email, password) {
    await this.openModal();
    const hasToggleReg = await this.isElementPresent(this.toggleRegisterButton, 1000);
    if (hasToggleReg) {
      await this.click(this.toggleRegisterButton);
      await this.sleep(200);
    }

    await this.type(this.registerNameInput, fullName);
    await this.type(this.emailInput, email);
    await this.type(this.passwordInput, password);
    await this.click(this.submitButton);
    await this.sleep(600);
  }

  async isUserLoggedIn() {
    return await this.isElementPresent(this.userMenuButton, 4000);
  }

  async isErrorDisplayed() {
    return await this.isElementPresent(this.errorAlert, 3000);
  }

  async getErrorMessage() {
    if (await this.isErrorDisplayed()) {
      return await this.getText(this.errorAlert);
    }
    return '';
  }

  async logout() {
    await this.driver.executeScript(`
      try {
        localStorage.removeItem('novamart_token');
        const signout = document.querySelector("[data-testid='sign-out-btn']");
        if (signout) signout.click();
      } catch (e) {}
    `);
    await this.sleep(800);
    // If signin-button is still not there, refresh
    const signin = await this.isElementPresent(By.css("[data-testid='signin-button']"), 500);
    if (!signin) {
      await this.driver.navigate().refresh();
      await this.sleep(800);
    }
  }
}

File: automation/Pages/LoginPage.cs
using System;
using OpenQA.Selenium;
using NOVAMart.Automation.Configuration;
using NOVAMart.Automation.Utilities;

namespace NOVAMart.Automation.Pages
{
    public class LoginPage : BasePage
    {
        public LoginPage(IWebDriver driver) : base(driver) { }

        public IWebElement EmailInput =>
            WaitHelper.WaitForElementVisible(Driver, By.CssSelector("[data-testid='login-email']"));

        public IWebElement PasswordInput =>
            WaitHelper.WaitForElementVisible(Driver, By.CssSelector("[data-testid='login-password']"));

        public IWebElement SubmitButton =>
            WaitHelper.WaitForElementClickable(Driver, By.CssSelector("[data-testid='login-submit']"));

        public IWebElement CloseButton =>
            WaitHelper.WaitForElementClickable(Driver, By.CssSelector("[data-testid='login-modal-close']"));

        public IWebElement SwitchToRegisterButton =>
            WaitHelper.WaitForElementClickable(Driver, By.CssSelector("[data-testid='toggle-register-btn']"));

        public void OpenLoginModal()
        {
            if (IsUserLoggedIn())
            {
                Logout();
            }

            var modalInputs = Driver.FindElements(By.CssSelector("[data-testid='login-email']"));
            if (modalInputs.Count == 0 || !modalInputs[0].Displayed)
            {
                var signinButtons = Driver.FindElements(By.CssSelector("[data-testid='signin-button'], #navbar-signin-button"));
                if (signinButtons.Count > 0 && signinButtons[0].Displayed)
                {
                    signinButtons[0].Click();
                }
                else
                {
                    var js = (IJavaScriptExecutor)Driver;
                    js.ExecuteScript("document.querySelector('[data-testid=\\\"signin-button\\\"]')?.click();");
                }
            }

            WaitHelper.WaitForElementVisible(Driver, By.CssSelector("[data-testid='login-email']"), 8);
        }

        public void Login(string email, string password)
        {
            OpenLoginModal();

            // If register mode is currently showing, switch to login
            var toggleLoginButtons = Driver.FindElements(By.CssSelector("[data-testid='toggle-login-btn']"));
            if (toggleLoginButtons.Count > 0 && toggleLoginButtons[0].Displayed)
            {
                toggleLoginButtons[0].Click();
            }

            var emailField = EmailInput;
            emailField.Clear();
            if (!string.IsNullOrEmpty(email))
            {
                emailField.SendKeys(email);
            }

            var passwordField = PasswordInput;
            passwordField.Clear();
            if (!string.IsNullOrEmpty(password))
            {
                passwordField.SendKeys(password);
            }

            SubmitButton.Click();
        }

        public bool IsErrorDisplayed()
        {
            try
            {
                var alert = WaitHelper.WaitForElementVisible(Driver, By.CssSelector("[data-testid='auth-error-alert'], [data-testid='login-error'], .text-rose-600"), 4);
                return alert != null && alert.Displayed;
            }
            catch
            {
                return false;
            }
        }

        public string GetErrorMessage()
        {
            try
            {
                var alert = WaitHelper.WaitForElementVisible(Driver, By.CssSelector("[data-testid='auth-error-alert'], [data-testid='login-error'], .text-rose-600"), 4);
                return alert?.Text.Trim() ?? string.Empty;
            }
            catch
            {
                return string.Empty;
            }
        }

        public void CloseModal()
        {
            try
            {
                CloseButton.Click();
            }
            catch
            {
                var js = (IJavaScriptExecutor)Driver;
                js.ExecuteScript("document.querySelector('[data-testid=\\\"login-modal-close\\\"]')?.click();");
            }
        }

        public void SwitchToRegister()
        {
            OpenLoginModal();
            SwitchToRegisterButton.Click();
            WaitHelper.WaitForElementVisible(Driver, By.CssSelector("[data-testid='register-name']"));
        }
    }
}
